incubator/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSubscribeFactory.java

```diff
@@ -903,15 +1150,406 @@ private void doKafkaWindow(
+            BeginFW begin)
+        {
+            final long sequence = begin.sequence();
+            final long acknowledge = begin.acknowledge();
+            final int maximum = begin.maximum();
+            final long traceId = begin.traceId();
+            final long authorization = begin.authorization();
+            final long affinity = begin.affinity();
+
+            assert acknowledge <= sequence;
+            assert sequence >= replySeq;
+            assert acknowledge >= replyAck;
+
+            replySeq = replySeq + sequence;
+            replyAck = replySeq + acknowledge;
+            replyMax = maximum;
+
+            state = MqttKafkaState.openingReply(state);
+
+            assert replyAck <= replySeq;
+
+            mqtt.doMqttBegin(traceId, authorization, affinity);
+            doKafkaWindow(traceId, authorization, mqtt.replyBud, mqtt.replyPad, 0);
+        }
+
+        private void onKafkaData(
+            DataFW data)
+        {
+            final long sequence = data.sequence();
+            final long acknowledge = data.acknowledge();
+            final long traceId = data.traceId();
+            final long authorization = data.authorization();
+            final long budgetId = data.budgetId();
+            final int reserved = data.reserved();
+
+            assert acknowledge <= sequence;
+            assert sequence >= replySeq;
+
+            replySeq = sequence + reserved;
+
+            assert replyAck <= replySeq;
+
+            if (replySeq > replyAck + replyMax)
+            {
+                doKafkaReset(traceId);
+                mqtt.doMqttAbort(traceId, authorization);
+            }
+            else
+            {
+                final int flags = data.flags();
+                final int length = data.length();
+                final OctetsFW payload = data.payload();
+                final OctetsFW extension = data.extension();
+                final ExtensionFW dataEx = extension.get(extensionRO::tryWrap);
+                final KafkaDataExFW kafkaDataEx =
+                    dataEx != null && dataEx.typeId() == kafkaTypeId ? extension.get(kafkaDataExRO::tryWrap) : null;
+                final KafkaMergedDataExFW kafkaMergedDataEx =
+                    kafkaDataEx != null && kafkaDataEx.kind() == KafkaDataExFW.KIND_MERGED ? kafkaDataEx.merged() : null;
+                final OctetsFW key = kafkaMergedDataEx != null ? kafkaMergedDataEx.key().value() : null;
+                final long filters = kafkaMergedDataEx != null ? kafkaMergedDataEx.filters() : 0;
+
+                if (key != null)
+                {
+                    String topicName = kafkaMergedDataEx.key().value()
+                        .get((b, o, m) -> b.getStringWithoutLengthUtf8(o, m - o));
+                    helper.visit(kafkaMergedDataEx);
+                    final Flyweight mqttSubscribeDataEx = mqtt.createSubscribeDataEx(kafkaMergedDataEx, filters, topicName,
+                        true, mqtt.retainedSubscriptionIds);
```

Comment:
This is still using the "createXX" approach we discussed removing.

Reply:
Yep, I did not push the changes yet. Once I address a comment locally, I immediately mark it as resolved, so I have a clean overview what still needs to be done.